services/supervisor/src/transactions.rs

```diff
@@ -89,6 +120,127 @@ impl ValidateInput for StartService {
 }
 
 impl SupervisorInterface for Supervisor {
+    fn propose_config_change(
+        &self,
+        context: TransactionContext,
+        propose: ConfigPropose,
+    ) -> Result<(), ExecutionError> {
+        let ((_, author), fork) = context
+            .verify_caller(Caller::as_transaction)
+            .ok_or(DispatcherError::UnauthorizedCaller)?;
+        let schema = Schema::new(context.instance.name, fork);
+
+        // Verifies that transaction author is validator.
+        let mut config_confirms = schema.config_confirms();
+        config_confirms
+            .validator_id(author)
+            .ok_or(Error::UnknownAuthor)?;
+
+        // Verifies that the `actual_from` height is in the future.
+        if blockchain::Schema::new(fork).height() >= propose.actual_from {
+            return Err(Error::ActualFromIsPast.into());
+        }
+
+        // Verifies that there are no pending config changes.
+        if schema.pending_proposal().exists() {
+            return Err(Error::ConfigProposeExists.into());
+        }
+
+        // To prevent multiple consensus change proposition in one request
+        let mut consensus_propose_added = false;
+        // To prevent multiple service change proposition in one request
+        let mut service_ids = HashSet::new();
+
+        // Perform config verification.
+        for change in &propose.changes {
+            match change {
+                ConfigChange::Consensus(config) => {
+                    if consensus_propose_added {
+                        trace!("Discarded multiple consensus change propositions in one request.");
```

Comment:
Nit (here and in the other places): "proposition"/"proposal" mixed terminology. Let's be consistent.
However, that's absolutely not critical and can be ignored if you feel tired of nitpicks :)